### ISSUE TITLE
ASS subtitles not shown on gbm if GUI resolution limit is active

### ISSUE THREAD

**Issue by HiassofT** [xbmc/xbmc#20085]
### Describe the bug
When playing a video with ass subtitles on gbm and GUI resolution limit is enabled and the display resolution is higher than that (eg 4k output with 1080p resolution limit) then the subtitles aren't shown / visible.

This happens both when playing 4k videos and HD videos upscaled to 4k.

SRT subtitles are displayed fine, ASS subtitles are displayed fine if GUI resolution limit is disabled or display output is less than or equal to the limit.

## Expected Behavior
ASS subtitles are shown with eg 4k output and 1080p GUI resolution limit.

## Actual Behavior
No subtitles are shown

### To Reproduce
1. Run kodi gbm (eg on RPi4) using a 4k TV
2. Configure display to 3840x2160 and set GUI resolution limit to 1080p
3. Play a video with ASS subtitles

### Debuglog
The debuglog can be found here: http://ix.io/3xem

Big Bug Bunny 1080 with some random ASS subtitle file (taken from here https://forum.libreelec.tv/thread/23951-libreelec-rpi4-9-95-2-does-not-reproduce-ass-subtitles/?postID=155349#post155349) first played at 4k output (no subs visible), then at HD (suibs shown)

## Additional context or screenshots (if appropriate)
Checking with a debugger it looks like ass rendering might use the wrong coordinate system / sizes when calculating margins.

When playing BBB at 4k with 1080p limit CDVDSubtitlesLibass::RenderImage calculated very large negative margins:
```
Thread 1 "kodi.bin" hit Breakpoint 2, CDVDSubtitlesLibass::RenderImage (this=0xedd15ce0, frameWidth=frameWidth@entry=1920, frameHeight=frameHeight@entry=1080, videoWidth=videoWidth@entry=3840, videoHeight=videoHeight@entry=2160, sourceWidth=sourceWidth@entry=1920, sourceHeight=sourceHeight@entry=1080, pts=6033333, useMargin=0, position=0, changes=changes@entry=0xffb47eb0) at ../xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:169
169       ass_set_aspect_ratio(m_renderer, dar, sar);
(gdb) p topmargin
$1 = -540
(gdb) p leftmargin
$2 = -960
(gdb) p useMargin
$3 = 0
(gdb) p position
$4 = 0
(gdb) p sar
$5 = 1.7777777777777777
(gdb) p dar
$6 = 1.7777777777777777
```

One level up, in OVERLAY::CRenderer::Convert, I see the following overlay coordinates:
```
(gdb) p *overlay
$8 = {_vptr.COverlay = 0xdfb8fc <vtable for OVERLAY::COverlayGlyphGL+8>, m_type = OVERLAY::COverlay::TYPE_NONE, m_align = OVERLAY::COverlay::ALIGN_VIDEO, m_pos = OVERLAY::COverlay::POSITION_RELATIVE, m_x = 0.25, m_y = 0.25, m_width = 0.5, m_height = 0.5}
```

When running at 4k without GUI limit (and subs are visible) the margins in RenderImage are 0 and overlay coordinates are 0/1
```
Thread 1 "kodi.bin" hit Breakpoint 1, CDVDSubtitlesLibass::RenderImage (this=0xf2e56958, frameWidth=frameWidth@entry=3840, frameHeight=frameHeight@entry=2160, videoWidth=videoWidth@entry=3840, videoHeight=videoHeight@entry=2160, sourceWidth=sourceWidth@entry=1920, sourceHeight=sourceHeight@entry=1080, pts=6000000, useMargin=0, position=0, changes=changes@entry=0xffc603c0) at ../xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:169
169       ass_set_aspect_ratio(m_renderer, dar, sar);
(gdb) p topmargin
$1 = 0
(gdb) p leftmargin
$2 = 0
(gdb) p sar
$3 = 1.7777777777777777
(gdb) p dar
$4 = 1.7777777777777777
(gdb) p useMargin
$5 = 0
(gdb) p position
$6 = 0
...
Thread 1 "kodi.bin" hit Breakpoint 2, OVERLAY::CRenderer::Convert (this=this@entry=0x26bfb78, o=o@entry=0xf2e3b9e0, pts=6000000) at ../xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp:414
414         overlay->m_y = ((float)videoHeight - targetHeight) / 2 / videoHeight;
(gdb) n
416       m_textureCache[m_textureid] = overlay;
(gdb) p *overlay
$8 = {_vptr.COverlay = 0xdfb8fc <vtable for OVERLAY::COverlayGlyphGL+8>, m_type = OVERLAY::COverlay::TYPE_NONE, m_align = OVERLAY::COverlay::ALIGN_VIDEO, m_pos = OVERLAY::COverlay::POSITION_RELATIVE, m_x = 0, m_y = 0, m_width = 1, m_height = 1}
```

### Your Environment
Used Operating system:
 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [X] Linux
 - [ ] OSX
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: LibreELEC 10, Raspberry Pi 4
 - Kodi version: 19.1+, githash 6402a50950ff6463042f8891ec4029c782730391

**Comment by nickdagamer400** [comment]
As far as I know, the subtitles are shown, but are drawn off-screen. In videos where the ass subtitles are sometimes in the center of the screen, they appear larger and further down/to the right than normal. So it seems something with Kodi is miscalculating the positioning of the subtitles if the GUI resolution limit is active, rather than not drawing them whatsoever.

Pardon my lack of understanding of the technical details, but that's what I believe I can observe.

**Comment by CastagnaIT** [comment]
Hi i am currently working on Ass implementation but i have not found the mentioned setting
"set GUI resolution limit"
could be this setting a custom implemetation of LibreELEC ?

**Comment by fuzzard** [comment]
Its an android only setting

https://github.com/xbmc/xbmc/blob/master/system/settings/android.xml#L15-L28

**Comment by CastagnaIT** [comment]
yes the margins are also affected by the problem because the "frame" size is lower than the "video" size
but in your case, the margins values are unused (so disabled)

If I haven't calculated wrong, the problem of having a smaller frame
cause a wrong calculation of `state.x` `state.y`
in theory 960, 540 instead to have 0, 0
https://github.com/xbmc/xbmc/blob/19.1-Matrix/xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp#L290-L302

@HiassofT if you try run the video at 4k with GUI limit enabled, and important, set: `Subtitle position on screen` to `Bottom of video`
the SRT subtitles should be in wrong position or offscreen like ASS, please if you can let me know

**Comment by HiassofT** [comment]
@fuzzard the setting is linux/GBM specific (didn't know android had a similar one) https://github.com/xbmc/xbmc/blob/master/system/settings/linux.xml#L199-L213 - in LE we enable it on ARM devices via appliance.xml, eg https://github.com/xbmc/xbmc/blob/master/system/settings/linux.xml#L199-L213

@CastagnaIT yes, I can confirm playing BBB 1080 with a srt file at 4k with 1080 gui limit doesn't show the sub, too, if the position is set to Bottom of Video (probably off-screen). Setting position to fixed or display to 1080p makes the sub visible
